ts/import-csv/index.ts

```diff
@@ -0,0 +1,83 @@
+// Copyright: Ankitects Pty Ltd and contributors
+// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
+
+import "./import-csv-base.css";
+
+import { ModuleName, setupI18n } from "../lib/i18n";
+import { checkNightMode } from "../lib/nightmode";
+import {
+    Decks,
+    decks as decksService,
+    empty,
+    notetypes as notetypeService,
+} from "../lib/proto";
+import ImportCsvPage from "./ImportCsvPage.svelte";
+import { getCsvMetadata } from "./lib";
+
+const gettingNotetypes = notetypeService.getNotetypeNames(empty);
+const gettingDecks = decksService.getDeckNames(
+    Decks.GetDeckNamesRequest.create({
+        skipEmptyDefault: false,
+        includeFiltered: false,
+    }),
+);
+const i18n = setupI18n({
+    modules: [
+        ModuleName.ACTIONS,
+        ModuleName.CHANGE_NOTETYPE,
+        ModuleName.DECKS,
+        ModuleName.EDITING,
+        ModuleName.IMPORTING,
+        ModuleName.KEYBOARD,
+        ModuleName.NOTETYPES,
+        ModuleName.STUDYING,
+    ],
+});
+
+export async function setupImportCsvPage(path: string): Promise<ImportCsvPage> {
+    const gettingMetadata = getCsvMetadata(path);
+    const [notetypes, decks, metadata] = await Promise.all([
+        gettingNotetypes,
+        gettingDecks,
+        gettingMetadata,
+        i18n,
+    ]);
+
+    const [deckId, deckColumn] =
```

Comment:
Ugh. Is there a better way to deal with oneofs in TS?
One idea I had was to use one-based column indices. Then `deckColumn == 0` would always mean unset, instead of now where it can mean either unset or set to 0.

Apart from that, should we use one-based indices in the CSV file, anyway? I.e. `#deck column:1` for the first column in the file.

Reply:
Like in Python, I think we can access the fields directly and rely on them being null/undefined in the case that it's a different oneof, eg something like:

```typescript
    const deckId = metadata.deckId ?? null;
    const deckColumn = metadata.deckColumn ?? null;
```

Or if you wanted to enforce deckColumn being null if deckId is defined, something like:

```typescript
    const deckColumn = deckId === null ? metadata.deckColumn ?? null : null;
```

1-based indexing is the more user-friendly option, and deferring the translation to 0-based indexing until the last moment could potentially save us having to deal with oneofs. It does increase the risk of off-by-one bugs - if not in the initial implementation, then when we return to the code at a later date and have forgotten how we chose to implement it. :-) But either approach seems reasonable, and I don't have a strong preference.

Reply:
> Like in Python, I think we can access the fields directly and rely on them being null/undefined in the case that it's a different oneof

Don't we use `HasField()` in Python because that's exactly not so and unset fields yield a default value?
I ran into that exact issue here: `metadata.deckColumn` is 0 if unset.

> It does increase the risk of off-by-one bugs

I agree, hopefully it could be minimised with tests and comments. Since we label columns starting with "Column 1" in the GUI, I think it would be really confusing to expect zero-based indices in the file.

Reply:
Ah, you're right. Not the first time that behaviour has tripped me up. Life would be easier if we only had to deal with tagged unions in Rust!

> I think it would be really confusing to expect zero-based indices in the file.

:+1: